integration_test/tests/free_and_cancel_tests.rs

```diff
@@ -0,0 +1,142 @@
+                cstr::to_widechar_vec("SELECT * FROM integration_test.foo");
+            query.push(0);
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLPrepareW(stmt_handle, query.as_ptr(), SQL_NTS as i32),
+            );
+
+            get_column_attributes(
+                stmt_handle as Handle,
+                2,
+                Some(vec![Desc::SQL_DESC_UNSIGNED, Desc::SQL_DESC_UPDATABLE]),
+                true,
+            );
+
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_handle, FreeStmtOption::SQL_CLOSE)
+            );
+
+            disconnect_and_free_dbc_and_env_handles(env_handle, conn_handle);
+        }
+    }
+
+    /// This test is inspired by the SSIS Preview Data data retrieval flow.
+    /// It is altered to be more general than that specific flow, with a focus
+    /// on canceling the query after getting some data. This flow depends on
+    /// setup_connect_and_alloc_stmt.
+    /// After allocating a statement handle, the flow is:
+    ///     - SQLSetStmtAttrW(SQL_ATTR_QUERY_TIMEOUT, 15)
+    ///     - SQLExecDirectW(<query>)
+    ///     - SQLRowCount
+    ///     - <loop: for each column>
+    ///         - SQLColAttributeW(SQL_DESC_CONCISE_TYPE)
+    ///         - SQLColAttributeW(SQL_DESC_UNSIGNED)
+    ///         - SQLColAttributeW(SQL_COLUMN_NAME)
+    ///         - SQLColAttributeW(SQL_COLUMN_NULLABLE)
+    ///         - SQLColAttributeW(SQL_DESC_TYPE_NAME)
+    ///         - SQLColAttributeW(SQL_COLUMN_LENGTH)
+    ///         - SQLColAttributeW(SQL_COLUMN_SCALE)
+    ///     - <loop: until SQLFetch return SQL_NO_DATA>
+    ///         - SQLFetch
+    ///         - SQLGetData
+    ///     - SQLCancel
+    #[test]
+    fn test_cancel() {
+        let (_, _, stmt_handle) = setup_connect_and_alloc_stmt();
+
+        unsafe {
+            let timeout: i32 = 15;
+            let value = timeout as Pointer;
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLSetStmtAttrW(
+                    stmt_handle,
+                    StatementAttribute::SQL_ATTR_QUERY_TIMEOUT,
+                    value,
+                    0
+                )
+            );
+
+            let mut query: Vec<WideChar> =
+                cstr::to_widechar_vec("SELECT * FROM integration_test.foo");
+            query.push(0);
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(stmt_handle, query.as_ptr(), SQL_NTS as i32)
+            );
+
+            let row_count_ptr = &mut 0;
+            assert_eq!(SqlReturn::SUCCESS, SQLRowCount(stmt_handle, row_count_ptr));
+
+            get_column_attributes(stmt_handle as Handle, 2, None, false);
+
+            fetch_and_get_data(
+                stmt_handle as Handle,
+                Some(3),
+                vec![SqlReturn::SUCCESS; 2],
+                vec![CDataType::SQL_C_SSHORT, CDataType::SQL_C_SLONG],
```

Comment:
I think this call is right, but also seems like we don't support `SQL_C_SSHORT` in `SQLGetData`. i imagine it wouldn't be _hard_, and if I'm understanding the spec right it seems like we should? looking at [this](https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/c-data-types?view=sql-server-ver16) doc it would seem the types we dont support that we probably could are `SQL_C_SSHORT`, `SQL_C_USHORT`, `SQL_C_STINYINT` and `SQL_C_UTINYINT`; and I imagine the bookmark types we would ignore.

Wondering if either you or @nbagnard have any context on whether there was any reason not to include them, or not implement them now? (current location is [here](https://github.com/mongodb/mongo-odbc-driver/blob/43bb8a95c1a4b1946daa57f1abf8f4dbd29facb3/odbc/src/api/data.rs#L788))

Reply:
@jchemburkar 
If I remember correctly we didn't implement them because no BSON type maps to SQL_SMALLINT or SQL_TINYINT so we don't expect tools to want to get the data as C_xxSHORT or SQL_C_xxTINYINT because they would risk truncation in doing that.
We can support this yes and if tools need it, then we should.
As you figured out we "simply" need to add the support for the SQL_C_xxx type in the function you pointed to, map it to the correct rust type and then add to_i16, to_u16, etc....

Reply:
I looked at the mapping, data in the `foo` collection are Bson int, Bson long, so you should get => SQL_LONG, SQL_BIGINT in the metadata => SQL_C_SLONG, SQL_C_SQL_C_SBIGINT  for the C default/direct mapping=> i32, i64 as Rust type.